### PR TITLE
Fix for issue 3143: Import entry from clipboard in different formats

### PR DESCRIPTION
**Pull request by 125m125**
see #3143
Use all importers to parse pasted entries. This allows the user to paste entries in different formats and not ony BibTex.

- [X] Change in CHANGELOG.md described
- [X] Tests created for changes
- [X] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by lynyus** [comment]
Hi @125m125, as I see you have some troubles with our checkstyle rules. I recommend you to use our IDE config files from https://github.com/JabRef/jabref/tree/master/config. Which IDE are you using? We recommend IntelliJ IDEA or Eclipse.
You can see if the tests a pass by calling `gradle check`. If you only want to check for the code style rules, you can call `gradle checkstyleMain`.

If you have any questions, feel free to ask them at slack: https://jabref.slack.com

**Comment by 125m125** [comment]
I am using Eclipse and just had to format a file. Because the saveAction is set to only reformat edited lines a missing newline in the imports wasn't corrected. Also thanks for pointing out gradle check. I somehow missed that command.

**Comment by Siedlerchr** [comment]
@125m125  If you use eclipse, just run` gradlew eclipse `to import the correct code formatting rules

File: src/main/java/org/jabref/logic/importer/ImportFormatReader.java [hunk 1]
```diff
@@ -215,4 +218,39 @@ public UnknownFormatImport importUnknownFormat(Path filePath) throws ImportExcep
 
         throw new ImportException(Localization.lang("Could not find a suitable import format."));
     }
+
+    public UnknownFormatImport importUnknownFormatFromString(String data) throws ImportException {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/ImportFormatReader.java]: As far as I can see, this is mostly a copy of `importUnknownFormat(Path filePath)` and I would like to have a solution without code duplication. Maybe something along the following lines works?
````java
private UnknownFormatImport importUnknownFormat(Function<Importer, ParserResult> importDatabase, Function<Importer, Boolean> isRecognizedFormat) {
    // original method with obvious changes imFo.importDatabase -> importDatabase.apply(imFo)
}

public UnknownFormatImport importUnknownFormat(Path filePath) {
    return importUnknownFormat(importer -> importer.importDatabase(filePath, importFormatPreferences.getEncoding(), importer -> importer .isRecognizedFormat(filePath, importFormatPreferences.getEncoding());
}

public UnknownFormatImport importUnknownFormat(String data) {
    return importUnknownFormat(importer -> importer.importDatabase(data), importer -> importer .isRecognizedFormat(data));
}
````

File: src/main/java/org/jabref/logic/importer/ImportFormatReader.java [hunk 2]
```diff
@@ -215,4 +218,39 @@ public UnknownFormatImport importUnknownFormat(Path filePath) throws ImportExcep
...
+        // stores ref to best result, gets updated at the next loop
+        List<BibEntry> bestResult = null;
+        long bestResultCount = 0;
+        String bestFormatName = null;
+
+        // Cycle through all importers:
+        for (Importer imFo : getImportFormats()) {
+            try (StringReader in = new StringReader(data); BufferedReader input = new BufferedReader(in)) {
+                ParserResult parserResult = imFo.importDatabase(input);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/ImportFormatReader.java]: Can you please extract the method `Importer.importDatabase(String data)`. This new method can be reused at least in the `MrDLibFetcher`. Similarly, for the `isRecognizedFormat`.

File: src/test/java/org/jabref/logic/importer/ImportFormatReaderIntegrationTest.java [hunk 3]
```diff
@@ -55,6 +55,20 @@ public void testImportFormatFromFile() throws Exception {
...
     }
 
+    @Test
+    public void testImportUnknownFormatFromString() throws Exception {
+        String data = new String(Files.readAllBytes(file), StandardCharsets.UTF_8);
+        try {
+            assertEquals(count, reader.importUnknownFormatFromString(data).parserResult.getDatabase().getEntries().size());
+        } catch (ImportException e) {
+            // msbib test file does not contain an author
+            if ("msbib".equals(format)) {
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/ImportFormatReaderIntegrationTest.java]: Why is this special treatment necessary for the import from a string, but not when the same content is imported from a file?

**Comment by 125m125** [comment]
Thanks for the feedback. I integrated the requests with the last commit.
I also changed the PdfXmpImporter to return false when checking the format instead of throwing an exception to prevent the need for special treatment and make it consistent with the PdfContentImporter.

**Comment by lenhard** [comment]
Ok, then I'll add this to the 4.1 milestone.

**Comment by tobiasdiez** [comment]
I think the error message comes from
https://github.com/JabRef/jabref/blob/ae6d01be46c96ef57e9871dfeb208ab0b7aab295/src/main/java/org/jabref/logic/importer/fileformat/MrDLibImporter.java#L53-L60
In my opinion, you can just delete these `Logger.log` statements (since its quite expected that a file in a wrong format is passed to `isRecognizedFormat` and this importer is not used outside of internal code anyways).

**Pull request merged**